src/Runtime/jni/jniwrapper.c

```diff
@@ -910,3 +918,26 @@ JNIEXPORT jstring JNICALL Java_com_ibm_onnxmlir_OMModel_output_1signature_1jni(
 
   return java_osig;
 }
+
+JNIEXPORT jobject JNICALL Java_com_ibm_onnxmlir_OMTensor_free_1tensor_1data(
+    JNIEnv *env, jclass cls, jobject jomt_data, jlong jomt_offset) {
+
+  log_init();
+
+  /* Find and initialize Java Exception class */
+  JNI_TYPE_VAR_CALL(env, jclass, jecpt_cls,
+      (*env)->FindClass(env, jnistr[CLS_JAVA_LANG_EXCEPTION]),
+      jecpt_cls != NULL, NULL, "Class java/lang/Exception not found");
+
+  /* Get direct buffer associated with data */
+  JNI_TYPE_VAR_CALL(env, void *, jni_data,
+      (*env)->GetDirectBufferAddress(env, jomt_data), jni_data != NULL,
+      jecpt_cls, "jni_data=%p", jni_data);
+
+  /* Free the data. By virtue of being here the data was allocated by the model
+     or the JNI and can be deleted. */
+  void *jomt_alloc = (void *)((char *)jni_data - (char *)jomt_offset);
```

Comment:
Instead of saving the `_offset` and needing to do the extra math to determine the `jomt_alloc` could we just save the allocation address?

Reply:
I thought about doing that but I wanted to avoid a limitation with java - all of the integral based arguments are signed. With an offset, which can be signed, I can keep the semantics more or less the same. With a `Cleaner` based approach I can create another DirectByteBuffer holding the actual allocation variable which the cleaner can use to clean up - no offset calculation required.